app/controllers/v3/service_offerings_controller.rb

```diff
@@ -1,37 +1,52 @@
 require 'fetchers/service_offerings_fetcher'
+require 'fetchers/service_plan_visibility_fetcher'
 require 'presenters/v3/service_offering_presenter'
 
 class ServiceOfferingsController < ApplicationController
   def show
     guid = hashed_params[:guid]
 
-    offering = if !current_user
-                 ServiceOfferingsFetcher.fetch_one_anonymously(guid)
-               elsif permission_queryer.can_read_globally?
-                 ServiceOfferingsFetcher.fetch_one(guid)
-               else
-                 ServiceOfferingsFetcher.fetch_one(guid, org_guids: permission_queryer.readable_org_guids)
-               end
+    not_authenticated! if !current_user && VCAP::CloudController::FeatureFlag.enabled?(:hide_marketplace_from_unauthenticated_users)
 
+    offering, space, public = ServiceOfferingsFetcher.fetch(guid)
     service_offering_not_found! if offering.nil?
 
-    presenter = Presenters::V3::ServiceOfferingPresenter.new(offering)
-    render status: :ok, json: presenter.to_json
+    if permission_queryer.can_read_globally? || public || visible_space_scoped?(space) || visible_in_readable_orgs?(offering)
+      presenter = Presenters::V3::ServiceOfferingPresenter.new(offering)
+      render status: :ok, json: presenter.to_json
+    else
+      service_offering_not_found!
+    end
   end
 
   def enforce_authentication?
-    return false if action_name == 'show'
-
-    super
+    action_name == 'show' ? false : super
   end
 
   def enforce_read_scope?
-    return false if action_name == 'show'
+    action_name == 'show' ? false : super
+  end
+
+  private
```

Comment:
this should be moved up to L#21 I think, since the `enforce_` methods are private in the application controller